src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java

```diff
@@ -1036,23 +1036,33 @@ public String toString() {
          */
         public List<Type> all_interfaces_field;
 
-        /* The 'other' projection: If 'this' is type of an inline class, then 'projection' is the
-           its doppleganger in its referene projection world and vice versa.
+        /** The 'other' projection: If 'this' is type of a primitive class, then 'projection' is the
+         *  reference projection type and vice versa. Lazily initialized, not to be accessed directly.
         */
         public ClassType projection;
 
+        /** Is this class type a reference projection of a primitive class type ?
+         */
+        private boolean isReferenceProjection;
+
         public ClassType(Type outer, List<Type> typarams, TypeSymbol tsym) {
-            this(outer, typarams, tsym, TypeMetadata.EMPTY);
+            this(outer, typarams, tsym, TypeMetadata.EMPTY, false);
         }
 
         public ClassType(Type outer, List<Type> typarams, TypeSymbol tsym,
                          TypeMetadata metadata) {
+            this(outer, typarams, tsym, metadata, false);
+        }
+
```

Comment:
I have rewired most of the constructor invocations to the new one which handles a reference projection flag parameter. A few call sites are still left in for the old constructors, these seem harmless.